### ISSUE TITLE
Code for generating training patches

### ISSUE THREAD

**Issue by vokcow** [iantsen/hecktor#3]
Dear Andrei,

Thanks for sharing your clean and well structured code! 

In the paper you say that, during training, patches are randomly extracted from the PET-CT images with probability 0.9 of containing the tumour class. However, the "make_dataset.py" file extracts the patches only according to the given bounding boxes (as during validation / inference). 

Please could you share a code snippet or give more detail of the patch extraction procedure during training? I have the following doubts:

- Is the 0.9 probability according to a uniform distribution? 
- Are the proportions of tumour with respect to background class controlled (e.g., with 0.9 probability patches will **_at least_** some percentage of tumour class?

Many thanks in advance,

Best wishes,

Víctor

**Comment by iantsen** [comment]
Hello @VokCow 

> In the paper you say that, during training, patches are randomly extracted from the PET-CT images with probability 0.9 of containing the tumour class. However, the "make_dataset.py" file extracts the patches only according to the given bounding boxes (as during validation / inference).

Yes, you are right. The script "make_dataset.py" uses the bounding boxes by default. I did not include the option to fully reproduce the training data, partly because my initial implementation was quite messy. However, you can do it this way:

In `make_dataset.py`:
1) Remove / comment the part with patch extraction (i.e., from `# Get bounding boxes:` to `# Resample all images using CT attributes:`) 
https://github.com/iantsen/hecktor/blob/5ebc6774d178139b52abebbe926b4cafabe3e29d/src/data/make_dataset.py#L39-L60

2) Since there is no cropping anymore, use proper names for all inputs of the resampling function `resample_sitk_image`, i.e., instead of `cr_img_ct` use `img_ct`, the same with pet and mask.

3) Replace https://github.com/iantsen/hecktor/blob/5ebc6774d178139b52abebbe926b4cafabe3e29d/src/data/make_dataset.py#L66
with `new_size=None` 

This will give you the resample data, which is not cropped using the provided bounding boxes. It means that for a single patient PET and CT scans will have the same size, but this size will differ among the patients. 

Therefore, PET/CT patches must be extracted during training. 
In `train.py`:
1) Add in `train_transforms`
https://github.com/iantsen/hecktor/blob/5ebc6774d178139b52abebbe926b4cafabe3e29d/model/train.py#L52-L57

the transform to extract the patch, namely `transforms.ExtractPatch(patch_size=(144, 144, 144), p_tumor=0.9)`. It is better if you add it **right after**
https://github.com/iantsen/hecktor/blob/5ebc6774d178139b52abebbe926b4cafabe3e29d/model/train.py#L53 

> Please could you share a code snippet or give more detail of the patch extraction procedure during training? I have the following doubts: 
    -     Is the 0.9 probability according to a uniform distribution?
    -     Are the proportions of tumour with respect to background class controlled (e.g., with 0.9 probability patches will at least some percentage of tumour class?

See `ExtractPatch` in `transforms.py` and the part of the answer written above.
Yes, it is the uniform distribution. With the prob 0.9 there is at least 1 tumor voxel, and the prob 0.1 corresponds to absolutely random patches (that might contain the tumor class too).

**Comment by vokcow** [comment]
Hi Andrei, 

Many thanks for your detailed answer! Indeed, I have missed that there is an `ExtractPatch` transform. With that and the other hints you gave I will be able to implement the training procedure hassle free. 

I will now close the issue. Best wishes,

Víctor
